File: python/src/ai_orchestration/ai_orchestrator.py
"""
AI Orchestrator Component
Central routing and coordination hub for all AI pipeline interactions.

Responsibilities:
- Analyze incoming patient requests to determine intent and routing
- Route requests to appropriate specialized pipelines
- Coordinate multi-pipeline requests for complex queries
- Monitor pipeline performance and implement circuit breakers
- Handle failover to backup systems when needed
"""

from datetime import datetime, timedelta
from enum import Enum
from typing import List, Dict, Any, Optional, Protocol
from pydantic import BaseModel, Field

from ..models.interaction_record import NormalizedRequest, PipelineRoute, PipelineResponse


class PipelineType(str, Enum):
    """Types of AI pipelines available"""
    CARE_STORY = "care_story"
    UNDERSTANDING_VERIFICATION = "understanding_verification"
    MEDICATION_CLARIFICATION = "medication_clarification"
    FOLLOW_UP = "follow_up"


class RequestAnalysis(BaseModel):
    """Analysis of a patient request for routing"""
    primary_intent: str
    secondary_intents: List[str]
    complexity: str = Field(..., pattern="^(simple|moderate|complex)$")
    requires_multiple_pipelines: bool
    suggested_pipelines: List[PipelineType]
    confidence: float = Field(..., ge=0.0, le=1.0)
    entities: Dict[str, Any]


class Pipeline(Protocol):
    """Protocol for AI pipeline implementations"""
    id: str
    name: str
    type: PipelineType
    is_available: bool
    average_response_time: float
    success_rate: float

    async def process(self, request: NormalizedRequest) -> PipelineResponse:
        """Process a request and return a response"""
        ...


class CoordinatedResponse(BaseModel):
    """Response from coordinated multi-pipeline execution"""
    responses: List[PipelineResponse]
    integrated_content: str
    confidence: float = Field(..., ge=0.0, le=1.0)
    sources: List[str]
    processing_time: float


class CircuitBreakerState(BaseModel):
    """State of a circuit breaker for a pipeline"""
    pipeline_id: str
    is_open: bool
    failure_count: int
    last_failure_time: Optional[datetime] = None
    next_retry_time: Optional[datetime] = None


class AIOrchestrator:
    """
    AI Orchestrator - Routes and coordinates AI pipeline requests
    """

    def __init__(
        self,
        failure_threshold: int = 3,
        circuit_breaker_timeout: int = 60,
    ):
        """
        Initialize the AI Orchestrator
        
        Args:
            failure_threshold: Number of failures before opening circuit breaker
            circuit_breaker_timeout: Timeout in seconds before retrying failed pipeline
        """
        self.pipelines: Dict[PipelineType, Pipeline] = {}
        self.circuit_breakers: Dict[str, CircuitBreakerState] = {}
        self.failure_threshold = failure_threshold
        self.circuit_breaker_timeout = circuit_breaker_timeout

    def register_pipeline(self, pipeline: Pipeline) -> None:
        """Register a pipeline with the orchestrator"""
        self.pipelines[pipeline.type] = pipeline
        self.circuit_breakers[pipeline.id] = CircuitBreakerState(
            pipeline_id=pipeline.id,
            is_open=False,
            failure_count=0,
        )

    def analyze_request(self, request: NormalizedRequest) -> RequestAnalysis:
        """
        Analyze incoming request to determine routing strategy
        Implements Requirement 7.1: Request analysis and routing
        """
        intent = request.intent.lower()
        entities = request.entities

        # Determine primary intent and suggested pipelines
        primary_intent = intent
        suggested_pipelines: List[PipelineType] = []
        complexity = "simple"
        confidence = 0.8

        # Care story generation requests
        if self._matches_intent(intent, ['care story', 'discharge', 'instructions', 'care plan']):
            primary_intent = 'care_story_generation'
            suggested_pipelines.append(PipelineType.CARE_STORY)

        # Understanding verification requests
        elif self._matches_intent(intent, ['confused', 'clarify', 'explain', 'understand', 'what does', 'help me understand']):
            primary_intent = 'understanding_verification'
            suggested_pipelines.append(PipelineType.UNDERSTANDING_VERIFICATION)

        # Medication-related requests
        elif self._matches_intent(intent, ['medication', 'medicine', 'drug', 'pill', 'prescription', 'dose']):
            primary_intent = 'medication_clarification'
            suggested_pipelines.append(PipelineType.MEDICATION_CLARIFICATION)

        # Follow-up and reminder requests
        elif self._matches_intent(intent, ['reminder', 'appointment', 'follow-up', 'schedule', 'when should']):
            primary_intent = 'follow_up'
            suggested_pipelines.append(PipelineType.FOLLOW_UP)

        # Default to understanding verification for unclear intents
        else:
            primary_intent = 'general_query'
            suggested_pipelines.append(PipelineType.UNDERSTANDING_VERIFICATION)
            confidence = 0.5

        # Check for multi-pipeline needs
        secondary_intents: List[str] = []

        # If medication mentioned in care story context
        if (PipelineType.CARE_STORY in suggested_pipelines and
                self._matches_intent(intent, ['medication', 'medicine'])):
            secondary_intents.append('medication_clarification')
            suggested_pipelines.append(PipelineType.MEDICATION_CLARIFICATION)
            complexity = 'moderate'

        # If follow-up mentioned with other intents
        if (self._matches_intent(intent, ['reminder', 'appointment']) and
                len(suggested_pipelines) > 0 and
                PipelineType.FOLLOW_UP not in suggested_pipelines):
            secondary_intents.append('follow_up')
            suggested_pipelines.append(PipelineType.FOLLOW_UP)
            complexity = 'moderate'

        # Complex queries require multiple pipelines
        if len(suggested_pipelines) > 2:
            complexity = 'complex'

        return RequestAnalysis(
            primary_intent=primary_intent,
            secondary_intents=secondary_intents,
            complexity=complexity,
            requires_multiple_pipelines=len(suggested_pipelines) > 1,
            suggested_pipelines=suggested_pipelines,
            confidence=confidence,
            entities=entities,
        )

    async def route_to_pipeline(
        self,
        analysis: RequestAnalysis,
        request: NormalizedRequest,
    ) -> List[PipelineRoute]:
        """
        Route request to appropriate pipeline(s)
        Implements Requirement 7.2: Pipeline selection and routing
        """
        routes: List[PipelineRoute] = []

        # Sort pipelines by priority (primary intent first)
        sorted_pipelines = self._prioritize_pipelines(analysis.suggested_pipelines)

        for i, pipeline_type in enumerate(sorted_pipelines):
            pipeline = self.pipelines.get(pipeline_type)

            if not pipeline:
                continue

            # Check circuit breaker
            circuit_breaker = self.circuit_breakers.get(pipeline.id)
            if circuit_breaker and self._is_circuit_breaker_open(circuit_breaker):
                # Try to find backup pipeline
                continue

            routes.append(PipelineRoute(
                pipeline_id=pipeline.id,
                pipeline_name=pipeline.name,
                routing_reason='primary_intent' if i == 0 else 'secondary_intent',
                priority=i + 1,
                timestamp=datetime.now(),
            ))

        return routes

    async def coordinate_multi_pipeline(
        self,
        pipelines: List[Pipeline],
        request: NormalizedRequest,
    ) -> CoordinatedResponse:
        """
        Coordinate multiple pipelines for complex requests
        Implements Requirement 7.3: Multi-pipeline coordination
        """
        import asyncio
        import time

        start_time = time.time()
        responses: List[PipelineResponse] = []
        sources: set = set()

        # Execute pipelines in parallel for efficiency
        async def execute_pipeline(pipeline: Pipeline) -> Optional[PipelineResponse]:
            try:
                response = await self._execute_pipeline_with_circuit_breaker(pipeline, request)
                responses.append(response)

                # Collect sources
                if response.sources:
                    for source in response.sources:
                        sources.add(source)

                return response
            except Exception as e:
                # Log error but continue with other pipelines
                print(f"Pipeline {pipeline.name} failed: {e}")
                return None

        # Execute all pipelines concurrently
        await asyncio.gather(*[execute_pipeline(p) for p in pipelines])

        # Filter out failed responses
        successful_responses = [r for r in responses if r is not None]

        # Integrate responses
        integrated_content = self._integrate_responses(successful_responses)
        average_confidence = (
            sum(r.confidence for r in successful_responses) / len(successful_responses)
            if successful_responses else 0.0
        )

        processing_time = time.time() - start_time

        return CoordinatedResponse(
            responses=successful_responses,
            integrated_content=integrated_content,
            confidence=average_confidence,
            sources=list(sources),
            processing_time=processing_time,
        )

    async def handle_failover(
        self,
        failed_pipeline: Pipeline,
        request: NormalizedRequest,
    ) -> PipelineResponse:
        """
        Handle pipeline failover to backup systems
        Implements Requirement 7.5: Failover and backup routing
        """
        # Mark circuit breaker
        self._record_failure(failed_pipeline.id)

        # Try to find alternative pipeline of same type
        alternative_pipeline = self._find_alternative_pipeline(failed_pipeline.type)

        if alternative_pipeline:
            try:
                return await alternative_pipeline.process(request)
            except Exception:
                # Alternative also failed
                self._record_failure(alternative_pipeline.id)

        # Return fallback response
        return PipelineResponse(
            pipeline_id='fallback',
            pipeline_name='Fallback Handler',
            response_content=(
                "I apologize, but I'm having trouble processing your request right now. "
                "Please try again in a moment, or contact your care team for immediate assistance."
            ),
            confidence=0.3,
            processing_time=0.0,
            timestamp=datetime.now(),
        )

    async def _execute_pipeline_with_circuit_breaker(
        self,
        pipeline: Pipeline,
        request: NormalizedRequest,
    ) -> PipelineResponse:
        """Execute pipeline with circuit breaker protection"""
        circuit_breaker = self.circuit_breakers.get(pipeline.id)

        if circuit_breaker and self._is_circuit_breaker_open(circuit_breaker):
            raise Exception(f"Circuit breaker open for pipeline {pipeline.id}")

        try:
            response = await pipeline.process(request)
            self._record_success(pipeline.id)
            return response
        except Exception as e:
            self._record_failure(pipeline.id)
            raise e

    def _is_circuit_breaker_open(self, circuit_breaker: CircuitBreakerState) -> bool:
        """Check if circuit breaker is open"""
        if not circuit_breaker.is_open:
            return False

        # Check if timeout has passed
        if circuit_breaker.next_retry_time and datetime.now() >= circuit_breaker.next_retry_time:
            # Reset circuit breaker
            circuit_breaker.is_open = False
            circuit_breaker.failure_count = 0
            return False

        return True

    def _record_success(self, pipeline_id: str) -> None:
        """Record pipeline success"""
        circuit_breaker = self.circuit_breakers.get(pipeline_id)
        if circuit_breaker:
            circuit_breaker.failure_count = 0
            circuit_breaker.is_open = False

    def _record_failure(self, pipeline_id: str) -> None:
        """Record pipeline failure and potentially open circuit breaker"""
        circuit_breaker = self.circuit_breakers.get(pipeline_id)
        if not circuit_breaker:
            return

        circuit_breaker.failure_count += 1
        circuit_breaker.last_failure_time = datetime.now()

        if circuit_breaker.failure_count >= self.failure_threshold:
            circuit_breaker.is_open = True
            circuit_breaker.next_retry_time = datetime.now() + timedelta(
                seconds=self.circuit_breaker_timeout
            )

    def _find_alternative_pipeline(self, pipeline_type: PipelineType) -> Optional[Pipeline]:
        """Find alternative pipeline of the same type"""
        # In a real implementation, this would check for backup pipelines
        # For now, return None as we don't have backup pipelines configured
        return None

    def _integrate_responses(self, responses: List[PipelineResponse]) -> str:
        """Integrate multiple pipeline responses into coherent output"""
        if not responses:
            return ''

        if len(responses) == 1:
            return responses[0].response_content

        # For multiple responses, combine them with clear sections
        sections = [response.response_content for response in responses]
        return '\n\n'.join(sections)

    def _prioritize_pipelines(self, pipelines: List[PipelineType]) -> List[PipelineType]:
        """Prioritize pipelines based on analysis"""
        # Primary intent pipeline should be first
        # For now, just return as-is since they're already ordered by priority
        return list(pipelines)

    def _matches_intent(self, intent: str, keywords: List[str]) -> bool:
        """Check if intent matches any of the keywords"""
        lower_intent = intent.lower()
        return any(keyword.lower() in lower_intent for keyword in keywords)

    def get_performance_metrics(self) -> Dict[str, Any]:
        """
        Get performance metrics for monitoring
        Implements Requirement 7.5: Performance monitoring
        """
        metrics: Dict[str, Any] = {}

        for pipeline_type, pipeline in self.pipelines.items():
            circuit_breaker = self.circuit_breakers.get(pipeline.id)
            metrics[pipeline_type.value] = {
                'is_available': pipeline.is_available,
                'average_response_time': pipeline.average_response_time,
                'success_rate': pipeline.success_rate,
                'circuit_breaker_open': circuit_breaker.is_open if circuit_breaker else False,
                'failure_count': circuit_breaker.failure_count if circuit_breaker else 0,
            }

        return metrics

File: src/ai-orchestration-layer/ai-orchestrator.ts
/**
 * AI Orchestrator Component
 * Central routing and coordination hub for all AI pipeline interactions.
 * 
 * Responsibilities:
 * - Analyze incoming patient requests to determine intent and routing
 * - Route requests to appropriate specialized pipelines
 * - Coordinate multi-pipeline requests for complex queries
 * - Monitor pipeline performance and implement circuit breakers
 * - Handle failover to backup systems when needed
 */

import {
  NormalizedRequest,
  PipelineRoute,
  PipelineResponse,
} from '../shared/types/interaction-record';

export enum PipelineType {
  CARE_STORY = 'care_story',
  UNDERSTANDING_VERIFICATION = 'understanding_verification',
  MEDICATION_CLARIFICATION = 'medication_clarification',
  FOLLOW_UP = 'follow_up',
}

export interface RequestAnalysis {
  primaryIntent: string;
  secondaryIntents: string[];
  complexity: 'simple' | 'moderate' | 'complex';
  requiresMultiplePipelines: boolean;
  suggestedPipelines: PipelineType[];
  confidence: number;
  entities: Record<string, any>;
}

export interface Pipeline {
  id: string;
  name: string;
  type: PipelineType;
  isAvailable: boolean;
  averageResponseTime: number;
  successRate: number;
  process(request: NormalizedRequest): Promise<PipelineResponse>;
}

export interface CoordinatedResponse {
  responses: PipelineResponse[];
  integratedContent: string;
  confidence: number;
  sources: string[];
  processingTime: number;
}

export interface CircuitBreakerState {
  pipelineId: string;
  isOpen: boolean;
  failureCount: number;
  lastFailureTime?: Date;
  nextRetryTime?: Date;
}

/**
 * AI Orchestrator - Routes and coordinates AI pipeline requests
 */
export class AIOrchestrator {
  private pipelines: Map<PipelineType, Pipeline>;
  private circuitBreakers: Map<string, CircuitBreakerState>;
  private readonly failureThreshold: number = 3;
  private readonly circuitBreakerTimeout: number = 60000; // 60 seconds

  constructor() {
    this.pipelines = new Map();
    this.circuitBreakers = new Map();
  }

  /**
   * Register a pipeline with the orchestrator
   */
  registerPipeline(pipeline: Pipeline): void {
    this.pipelines.set(pipeline.type, pipeline);
    this.circuitBreakers.set(pipeline.id, {
      pipelineId: pipeline.id,
      isOpen: false,
      failureCount: 0,
    });
  }

  /**
   * Analyze incoming request to determine routing strategy
   * Implements Requirement 7.1: Request analysis and routing
   */
  analyzeRequest(request: NormalizedRequest): RequestAnalysis {
    const intent = request.intent.toLowerCase();
    const entities = request.entities;
    
    // Determine primary intent and suggested pipelines
    let primaryIntent = intent;
    let suggestedPipelines: PipelineType[] = [];
    let complexity: 'simple' | 'moderate' | 'complex' = 'simple';
    let confidence = 0.8;

    // Care story generation requests
    if (this.matchesIntent(intent, ['care story', 'discharge', 'instructions', 'care plan'])) {
      primaryIntent = 'care_story_generation';
      suggestedPipelines.push(PipelineType.CARE_STORY);
    }
    
    // Understanding verification requests
    else if (this.matchesIntent(intent, ['confused', 'clarify', 'explain', 'understand', 'what does', 'help me understand'])) {
      primaryIntent = 'understanding_verification';
      suggestedPipelines.push(PipelineType.UNDERSTANDING_VERIFICATION);
    }
    
    // Medication-related requests
    else if (this.matchesIntent(intent, ['medication', 'medicine', 'drug', 'pill', 'prescription', 'dose'])) {
      primaryIntent = 'medication_clarification';
      suggestedPipelines.push(PipelineType.MEDICATION_CLARIFICATION);
    }
    
    // Follow-up and reminder requests
    else if (this.matchesIntent(intent, ['reminder', 'appointment', 'follow-up', 'schedule', 'when should'])) {
      primaryIntent = 'follow_up';
      suggestedPipelines.push(PipelineType.FOLLOW_UP);
    }
    
    // Default to understanding verification for unclear intents
    else {
      primaryIntent = 'general_query';
      suggestedPipelines.push(PipelineType.UNDERSTANDING_VERIFICATION);
      confidence = 0.5;
    }

    // Check for multi-pipeline needs
    const secondaryIntents: string[] = [];
    
    // If medication mentioned in care story context
    if (suggestedPipelines.includes(PipelineType.CARE_STORY) && 
        this.matchesIntent(intent, ['medication', 'medicine'])) {
      secondaryIntents.push('medication_clarification');
      suggestedPipelines.push(PipelineType.MEDICATION_CLARIFICATION);
      complexity = 'moderate';
    }
    
    // If follow-up mentioned with other intents
    if (this.matchesIntent(intent, ['reminder', 'appointment']) && 
        suggestedPipelines.length > 0 && 
        !suggestedPipelines.includes(PipelineType.FOLLOW_UP)) {
      secondaryIntents.push('follow_up');
      suggestedPipelines.push(PipelineType.FOLLOW_UP);
      complexity = 'moderate';
    }

    // Complex queries require multiple pipelines
    if (suggestedPipelines.length > 2) {
      complexity = 'complex';
    }

    return {
      primaryIntent,
      secondaryIntents,
      complexity,
      requiresMultiplePipelines: suggestedPipelines.length > 1,
      suggestedPipelines,
      confidence,
      entities,
    };
  }

  /**
   * Route request to appropriate pipeline(s)
   * Implements Requirement 7.2: Pipeline selection and routing
   */
  async routeToPipeline(
    analysis: RequestAnalysis,
    request: NormalizedRequest
  ): Promise<PipelineRoute[]> {
    const routes: PipelineRoute[] = [];
    
    // Sort pipelines by priority (primary intent first)
    const sortedPipelines = this.prioritizePipelines(analysis.suggestedPipelines);
    
    for (let i = 0; i < sortedPipelines.length; i++) {
      const pipelineType = sortedPipelines[i];
      const pipeline = this.pipelines.get(pipelineType);
      
      if (!pipeline) {
        continue;
      }

      // Check circuit breaker
      const circuitBreaker = this.circuitBreakers.get(pipeline.id);
      if (circuitBreaker && this.isCircuitBreakerOpen(circuitBreaker)) {
        // Try to find backup pipeline
        continue;
      }

      routes.push({
        pipelineId: pipeline.id,
        pipelineName: pipeline.name,
        routingReason: i === 0 ? 'primary_intent' : 'secondary_intent',
        priority: i + 1,
        timestamp: new Date(),
      });
    }

    return routes;
  }

  /**
   * Coordinate multiple pipelines for complex requests
   * Implements Requirement 7.3: Multi-pipeline coordination
   */
  async coordinateMultiPipeline(
    pipelines: Pipeline[],
    request: NormalizedRequest
  ): Promise<CoordinatedResponse> {
    const startTime = Date.now();
    const responses: PipelineResponse[] = [];
    const sources: Set<string> = new Set();

    // Execute pipelines in parallel for efficiency
    const pipelinePromises = pipelines.map(async (pipeline) => {
      try {
        const response = await this.executePipelineWithCircuitBreaker(pipeline, request);
        responses.push(response);
        
        // Collect sources
        if (response.sources) {
          response.sources.forEach(source => sources.add(source));
        }
        
        return response;
      } catch (error) {
        // Log error but continue with other pipelines
        console.error(`Pipeline ${pipeline.name} failed:`, error);
        return null;
      }
    });

    await Promise.all(pipelinePromises);

    // Filter out failed responses
    const successfulResponses = responses.filter(r => r !== null);

    // Integrate responses
    const integratedContent = this.integrateResponses(successfulResponses);
    const averageConfidence = successfulResponses.length > 0
      ? successfulResponses.reduce((sum, r) => sum + r.confidence, 0) / successfulResponses.length
      : 0;

    const processingTime = Date.now() - startTime;

    return {
      responses: successfulResponses,
      integratedContent,
      confidence: averageConfidence,
      sources: Array.from(sources),
      processingTime,
    };
  }

  /**
   * Handle pipeline failover to backup systems
   * Implements Requirement 7.5: Failover and backup routing
   */
  async handleFailover(
    failedPipeline: Pipeline,
    request: NormalizedRequest
  ): Promise<PipelineResponse> {
    // Mark circuit breaker
    this.recordFailure(failedPipeline.id);

    // Try to find alternative pipeline of same type
    const alternativePipeline = this.findAlternativePipeline(failedPipeline.type);
    
    if (alternativePipeline) {
      try {
        return await alternativePipeline.process(request);
      } catch (error) {
        // Alternative also failed
        this.recordFailure(alternativePipeline.id);
      }
    }

    // Return fallback response
    return {
      pipelineId: 'fallback',
      pipelineName: 'Fallback Handler',
      responseContent: 'I apologize, but I\'m having trouble processing your request right now. Please try again in a moment, or contact your care team for immediate assistance.',
      confidence: 0.3,
      processingTime: 0,
      timestamp: new Date(),
    };
  }

  /**
   * Execute pipeline with circuit breaker protection
   */
  private async executePipelineWithCircuitBreaker(
    pipeline: Pipeline,
    request: NormalizedRequest
  ): Promise<PipelineResponse> {
    const circuitBreaker = this.circuitBreakers.get(pipeline.id);
    
    if (circuitBreaker && this.isCircuitBreakerOpen(circuitBreaker)) {
      throw new Error(`Circuit breaker open for pipeline ${pipeline.id}`);
    }

    try {
      const response = await pipeline.process(request);
      this.recordSuccess(pipeline.id);
      return response;
    } catch (error) {
      this.recordFailure(pipeline.id);
      throw error;
    }
  }

  /**
   * Check if circuit breaker is open
   */
  private isCircuitBreakerOpen(circuitBreaker: CircuitBreakerState): boolean {
    if (!circuitBreaker.isOpen) {
      return false;
    }

    // Check if timeout has passed
    if (circuitBreaker.nextRetryTime && new Date() >= circuitBreaker.nextRetryTime) {
      // Reset circuit breaker
      circuitBreaker.isOpen = false;
      circuitBreaker.failureCount = 0;
      return false;
    }

    return true;
  }

  /**
   * Record pipeline success
   */
  private recordSuccess(pipelineId: string): void {
    const circuitBreaker = this.circuitBreakers.get(pipelineId);
    if (circuitBreaker) {
      circuitBreaker.failureCount = 0;
      circuitBreaker.isOpen = false;
    }
  }

  /**
   * Record pipeline failure and potentially open circuit breaker
   */
  private recordFailure(pipelineId: string): void {
    const circuitBreaker = this.circuitBreakers.get(pipelineId);
    if (!circuitBreaker) {
      return;
    }

    circuitBreaker.failureCount++;
    circuitBreaker.lastFailureTime = new Date();

    if (circuitBreaker.failureCount >= this.failureThreshold) {
      circuitBreaker.isOpen = true;
      circuitBreaker.nextRetryTime = new Date(Date.now() + this.circuitBreakerTimeout);
    }
  }

  /**
   * Find alternative pipeline of the same type
   */
  private findAlternativePipeline(type: PipelineType): Pipeline | null {
    // In a real implementation, this would check for backup pipelines
    // For now, return null as we don't have backup pipelines configured
    return null;
  }

  /**
   * Integrate multiple pipeline responses into coherent output
   */
  private integrateResponses(responses: PipelineResponse[]): string {
    if (responses.length === 0) {
      return '';
    }

    if (responses.length === 1) {
      return responses[0].responseContent;
    }

    // For multiple responses, combine them with clear sections
    const sections = responses.map((response, index) => {
      return response.responseContent;
    });

    return sections.join('\n\n');
  }

  /**
   * Prioritize pipelines based on analysis
   */
  private prioritizePipelines(pipelines: PipelineType[]): PipelineType[] {
    // Primary intent pipeline should be first
    // For now, just return as-is since they're already ordered by priority
    return [...pipelines];
  }

  /**
   * Check if intent matches any of the keywords
   */
  private matchesIntent(intent: string, keywords: string[]): boolean {
    const lowerIntent = intent.toLowerCase();
    return keywords.some(keyword => lowerIntent.includes(keyword.toLowerCase()));
  }

  /**
   * Get performance metrics for monitoring
   * Implements Requirement 7.5: Performance monitoring
   */
  getPerformanceMetrics(): Record<string, any> {
    const metrics: Record<string, any> = {};

    this.pipelines.forEach((pipeline, type) => {
      const circuitBreaker = this.circuitBreakers.get(pipeline.id);
      metrics[type] = {
        isAvailable: pipeline.isAvailable,
        averageResponseTime: pipeline.averageResponseTime,
        successRate: pipeline.successRate,
        circuitBreakerOpen: circuitBreaker?.isOpen || false,
        failureCount: circuitBreaker?.failureCount || 0,
      };
    });

    return metrics;
  }
}
